crates/server/src/authority/catalog.rs

```diff
@@ -178,23 +196,97 @@ impl Catalog {
     /// Constructs a new Catalog
     pub fn new() -> Self {
         Self {
-            authorities: HashMap::new(),
+            authorities: vec![(None, HashMap::new())],
         }
     }
 
-    /// Insert or update a zone authority
+    /// Insert or update a zone authority for all clients.
     ///
     /// # Arguments
     ///
     /// * `name` - zone name, e.g. example.com.
     /// * `authority` - the zone data
     pub fn upsert(&mut self, name: LowerName, authority: Box<dyn AuthorityObject>) {
-        self.authorities.insert(name, authority);
+        for (ipnet, authorities) in self.authorities.iter_mut().rev() {
+            if ipnet.is_none() {
+                authorities.insert(name, authority);
+                return;
+            }
+        }
+
+        self.authorities.push((None, {
+            let mut hm = HashMap::with_capacity(1);
+            hm.insert(name, authority);
+            hm
+        }));
+    }
+
+    /// Insert or update a zone authority for an [`IpNet`] range of clients.
+    ///
+    /// # Arguments
+    ///
+    /// * `ipnet` - [`IpNet`] this authority is for, e.g. 10.5.0.0/16
+    /// * `name` - zone name, e.g. example.com.
+    /// * `authority` - the zone data
+    pub fn upsert_for_ipnet(
```

Comment:
I want to think about this change here and the additional functions. It seems we could maybe fix this a little differently by maybe passing an object that has these parameters rather than creating alternative function names?

We have the request object, but I think that has a bunch of optionals that makes it unwieldy... so maybe we need a lookup object that has non-optional data in it that could be used for the authority lookup?

Reply:
I don't know what is best for the crate. My first intention was to avoid any breaking change so the pull request would have a chance to be accepted. I still believe this is a niche feature and doesn't worth changing the function signature...